core/src/main/java/org/apache/calcite/rel/metadata/RelMdFunctionalDependency.java

```diff
@@ -64,212 +76,422 @@ protected RelMdFunctionalDependency() {}
+      ImmutableBitSet dependents = inputFd.getDependents();
+
+      // Only preserve if both determinants and dependents are within group columns
+      if (groupSet.contains(determinants) && groupSet.contains(dependents)) {
+        fdSet.addFD(determinants, dependents);
+      }
     }
 
-    if (rel instanceof TableScan) {
-      TableScan tableScan = (TableScan) rel;
-      RelOptTable table = tableScan.getTable();
-      List<ImmutableBitSet> keys = table.getKeys();
-      return keys != null
-          && keys.size() == 1
-          && keys.get(0).equals(ImmutableBitSet.of(column));
-    } else if (rel instanceof Join) {
-      Join join = (Join) rel;
-      // TODO Considering column mapping based on equality conditions in join
-      int leftFieldCnt = join.getLeft().getRowType().getFieldCount();
-      if (key < leftFieldCnt && column < leftFieldCnt) {
-        return mq.determines(join.getLeft(), key, column);
-      } else if (key >= leftFieldCnt && column >= leftFieldCnt) {
-        return mq.determines(join.getRight(), key - leftFieldCnt, column - leftFieldCnt);
+    // 2. Group keys determine all aggregate columns
+    if (!groupSet.isEmpty() && !rel.getAggCallList().isEmpty()) {
+      for (int i = rel.getGroupCount(); i < rel.getRowType().getFieldCount(); i++) {
+        fdSet.addFD(groupSet, ImmutableBitSet.of(i));
       }
-      return false;
-    } else if (rel instanceof Correlate) {
-      // TODO Support Correlate.
-      return false;
-    } else if (rel instanceof SetOp) {
-      // TODO Support SetOp
-      return false;
     }
 
-    return mq.determines(rel.getInput(0), key, column);
+    return fdSet;
   }
 
-  private static Boolean preCheck(RelNode rel, int key, int column) {
-    verifyIndex(rel, key, column);
+  private static FunctionalDependencySet getJoinFD(Join rel, RelMetadataQuery mq) {
+    FunctionalDependencySet leftFdSet = mq.getFunctionalDependencies(rel.getLeft());
+    FunctionalDependencySet rightFdSet = mq.getFunctionalDependencies(rel.getRight());
+
+    int leftFieldCount = rel.getLeft().getRowType().getFieldCount();
+    JoinRelType joinType = rel.getJoinType();
+
+    switch (joinType) {
+    case INNER:
+      // Inner join: preserve all FDs and derive cross-table dependencies
+      FunctionalDependencySet innerJoinFdSet
+          = leftFdSet.union(shiftFdSet(rightFdSet, leftFieldCount));
+      deriveTransitiveFDs(rel, innerJoinFdSet, leftFieldCount);
+      return innerJoinFdSet;
+    case LEFT:
+      // Left join: preserve left FDs, right FDs may be invalidated by NULLs
+      FunctionalDependencySet leftJoinFdSet = new FunctionalDependencySet(leftFdSet.getFDs());
+      deriveTransitiveFDs(rel, leftJoinFdSet, leftFieldCount);
+      return leftJoinFdSet;
+    case RIGHT:
+      // Right join: preserve right FDs, left FDs may be invalidated by NULLs
+      FunctionalDependencySet shiftedRightFdSet = shiftFdSet(rightFdSet, leftFieldCount);
```

Comment:
this is still not symmetric with left - the transitive part is missing.

Reply:
Here, the index needs to be shifted because the initially obtained fdSet starts with an index of 0. I'm not sure if that's what you were asking about?

Reply:
shifted yes, but where is the transitive part that is done on the LEFT join?

Reply:
Got it, I missed the fdSet in the left part, I should have added it.